libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1164 @@
+/* radare - LGPL - Copyright 2019 - MapleLeaf-X */
+#include <string.h>
+#include "windows_debug.h"
+#include <windows.h>
+#include <tlhelp32.h> // CreateToolhelp32Snapshot
+#include <psapi.h> // GetModuleFileNameEx, GetProcessImageFileName
+#include <tchar.h>
+
+#ifndef XSTATE_GSSE
+#define XSTATE_GSSE 2
+#endif
+
+#ifndef XSTATE_LEGACY_SSE
+#define XSTATE_LEGACY_SSE 1
+#endif
+
+#if !defined(XSTATE_MASK_GSSE)
+#define XSTATE_MASK_GSSE (1LLU << (XSTATE_GSSE))
+#endif
+
+#undef CONTEXT_XSTATE
+#if defined(_M_X64)
+#define CONTEXT_XSTATE                      (0x00100040)
+#else
+#define CONTEXT_XSTATE                      (0x00010040)
+#endif
+#define XSTATE_AVX                          (XSTATE_GSSE)
+#define XSTATE_MASK_AVX                     (XSTATE_MASK_GSSE)
+#ifndef CONTEXT_ALL
+#define CONTEXT_ALL 1048607
+#endif
+
+typedef struct {
+	// bool dbgpriv;
+	HANDLE ph;
+	bool debug;
+} RIOW32Dbg;
+
+static ut64 (WINAPI *w32_GetEnabledXStateFeatures)() = NULL;
+static BOOL (WINAPI *w32_InitializeContext)(PVOID, DWORD, PCONTEXT*, PDWORD) = NULL;
+static BOOL (WINAPI *w32_GetXStateFeaturesMask)(PCONTEXT Context, PDWORD64) = NULL;
+static PVOID(WINAPI *w32_LocateXStateFeature)(PCONTEXT Context, DWORD, PDWORD) = NULL;
+static BOOL (WINAPI *w32_SetXStateFeaturesMask)(PCONTEXT Context, DWORD64) = NULL;
+
+static NTSTATUS (WINAPI *w32_NtQueryInformationThread)(HANDLE, ULONG, PVOID, ULONG, PULONG) = NULL;
+
+bool setup_debug_privileges(bool b) {
+	HANDLE tok;
+	if (!OpenProcessToken (GetCurrentProcess (), TOKEN_ADJUST_PRIVILEGES | TOKEN_QUERY, &tok)) {
+		return false;
+	}
+	bool ret = false;
+	LUID luid;
+	if (LookupPrivilegeValue (NULL, SE_DEBUG_NAME, &luid)) {
+		TOKEN_PRIVILEGES tp;
+		tp.PrivilegeCount = 1;
+		tp.Privileges[0].Luid = luid;
+		tp.Privileges[0].Attributes = b ? SE_PRIVILEGE_ENABLED : 0;
+		if (AdjustTokenPrivileges (tok, FALSE, &tp, 0, NULL, NULL)) {
+			// TODO: handle ERROR_NOT_ALL_ASSIGNED
+			ret = GetLastError () == ERROR_SUCCESS;
+		}
+	}
+	CloseHandle (tok);
+	return ret;
+}
+
+int w32_init(RDebug *dbg) {
+	RIOW32Dbg *rio = dbg->user = R_NEW (RIOW32Dbg);
+	if (!rio) {
+		eprintf ("w32_init: failed to allocate memory\n");
+		return false;
+	}
+	setup_debug_privileges (true);
+	// rio->dbgpriv = setup_debug_privileges (true);
+	rio->ph = (HANDLE)NULL;
+	rio->debug = false;
+
+	HMODULE lib = GetModuleHandle ("kernel32"); //Always loaded
```

Comment:
check if null

Reply:
This should never actually fail due to how kernel32 is loaded by the loader on windows but I will add the check if you like.